experimental/Schemes/Tjurina.jl

```diff
@@ -0,0 +1,770 @@
+  return vector_space_dimension(tjurina_algebra(f))
+end
+
+@doc raw"""
+    tjurina_number(X::AffineScheme{<:Field,<:MPolyQuoRing})
+
+Return the global Tjurina number of the affine scheme `X`, if `X` is a hypersurface.
+Return -1 if infinite.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> f = x^3 - y^2;
+
+julia> X = AffineScheme(quo(R, ideal(R, x^3-y^2))[1])
+Spectrum
+  of quotient
+    of multivariate polynomial ring in 2 variables x, y
+      over rational field
+    by ideal (x^3 - y^2)
+
+julia> tjurina_number(X)
+2
+```
+"""
+function tjurina_number(X::AffineScheme{<:Field,<:MPolyQuoRing}) 
+  return vector_space_dimension(tjurina_algebra(X))
+end
+
+
+
+@doc raw"""
+    tjurina_number(X::HypersurfaceGerm)
+    tjurina_number(X::HypersurfaceGerm, k::Integer = 0)
+
+Return the `k`-th local Tjurina number of `(X,p)` at `p`. 
+By default computes the local Tjurina number (`k=0`) at `p`.
+Higher Tjurina numbers are of interest in positive characteristic. Return -1 if infinite.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> f = x^3 - y^2;
+
+julia> X = HypersurfaceGerm(AffineScheme(quo(R, ideal(R, f))[1]), [0, 0]);
+
+julia> tjurina_number(X)
+2
+```
+"""
+function tjurina_number(X::HypersurfaceGerm, k::Integer = 0)                                    #TODO: simplify after vec_dim is improved
+  # Fix for zero dimensional vector space
+  I = modulus(tjurina_algebra(X, k))
+  if base_ring(I)(1) in I
+    return 0
+  end
+  # Fix for infinite dimensional vector space
+  try
+    return vector_space_dimension(tjurina_algebra(X, k))
+  catch
+    return -1
+  end
+end
+
+
+
+@doc raw"""
+    tjurina_number(f::MPolyLocRingElem{<:Field, <:Any, <:Any, <:Any, <:MPolyComplementOfKPointIdeal})
+    tjurina_number(f::MPolyLocRingElem{<:Field, <:Any, <:Any, <:Any, <:MPolyComplementOfKPointIdeal}, k::Integer = 0)
+
+Return the `k`-th local Tjurina number of `(V(f),p)` at `p`. 
+By default computes the local Tjurina number (`k=0`) at `p`.
+Higher Tjurina numbers are of interest in positive characteristic. 
+Return -1 if infinite.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> L,_  = localization(R, complement_of_point_ideal(R, [0, 0]));
+
+julia> f = L(x*y*(x-1));
+
+julia> tjurina_number(f)
+1
+```
+"""
+function tjurina_number(f::MPolyLocRingElem{<:Field, <:Any, <:Any, <:Any, <:MPolyComplementOfKPointIdeal}, k::Integer = 0)
+  X = HypersurfaceGerm(quo(parent(f), ideal(parent(f), f))[1])
+  return tjurina_number(X, k)
+end
+
+################################################################################
+
+#####                                 Order                                #####
+
+################################################################################
+
+
+function _order(f::MPolyRingElem)                                                                                 #TODO:  export???
+  f != 0 || return -1
+  x = gens(parent(f))
+  n = length(x)
```

Comment:
No need to allocate `x` if you then don't use it...
```suggestion
  n = ngens(parent(f))
```
or equivalently
```suggestion
  n = nvars(parent(f))
```